crates/red_knot/src/check.rs

```diff
@@ -0,0 +1,45 @@
+use crate::files::{FileId, Files};
+use crate::Workspace;
+use parking_lot::lock_api::RwLockUpgradableReadGuard;
+use parking_lot::RwLock;
+use rustc_hash::FxHashMap;
+use std::collections::HashMap;
+
+/// The basic idea of Salsa is that you give it all inputs (the source contents) and
+/// you can then define queries on top of it.
+///
+/// That means we need to eagerly set the contents of all files that the analysis could reach, which means
```

Comment:
It looks like it is possible to have lazy inputs in salsa: https://salsa-rs.github.io/salsa/common_patterns/on_demand_inputs.html

Reply:
Yeah I saw that. Trying it right now. There's still the problem that we can never purge the files